### DISCUSSION TITLE
Modal doesn't close, even though property is false

### DISCUSSION THREAD

**Discussion by estevanmaito** [alpinejs/alpine#528] [4. General]
You can reproduce it here https://codepen.io/estevanmaito/pen/PoPvbBz

If you click 'Open Modal', it opens the modal, and traps the focus inside the modal, so you can TAB indefinitely inside it. If you click outside of it, or in the close button or icon, it closes.

But, if you press ESC, it executes the function `close`, `this.show` turns to `false`, the cleanup is run but the modal is still visible 🤔

It's my first time using Alpine, so I may be missing something very obvious 😬

**Comment by ryangjchandler** [comment]
Hi @estevanmaito, I'm on mobile so could only give it a quick look. 

You should be doing `this.trapCleanup = trapFocus(...)` inside of `open`, your current code will implicitly create a temporary variable and not actually assign the return value.

You'll also need to do `this.trapCleanup()` instead of just `trapCleanup()`. 

For what it's worth, you could use Alpine's ~`x-on:keydown.esc.window`~ **edit:** `x-on:keydown.escape.window`  to run your close function too. 

Let me know if that helps.

**Reply by estevanmaito** [reply]
It gives this error now, that's probably related to `handleEsc` being run in the context of `window` now.

```
Uncaught TypeError: this.trapCleanup is not a function
    at close (modal.html:167)
    at HTMLDivElement.handleKeyDown (modal.html:127)
```

But even if I remove the call to `this.trapCleanup`, `close` is executed on every way, but on ESC it's ignored.

**Reply by SimoTod** [reply]
The code is a bit overly complicated but yeah it's a scope issue. Your this in the close function depends on the context where you run the function. When you pass it to the global function, this becomes window and stop being your alpine data context. If you want to pass it like that, before the return statement in your Alpine data function, you can add `const self = this` and use self instead of this inside `close()`
However, I would consoder a simpler implementation because, on a long run, it will get hard to maintain.

**Reply by estevanmaito** [reply]
@SimoTod I actually got it solved. It was a typo (see accepted answer). But I would like to understand what would you make simpler? The HTML part or the functions?

Talking about the trap focus function, I think it really does the job the leanest way possible, but I'm not familiar to the Alpine organization and would be very helpful to understand it.

**Reply by ryangjchandler** [reply]
Are you using the trapFocus function in multiple parts of your application? If not, I'd make it part of the component function instead.

**Reply by SimoTod** [reply]
If I understand correctly, the Alpine handler works, not the original example. What I found a bit complicated in the first snippet was that you were defining functions inside a literal object (let's ignore alpine for a sec, it's more a javascript issue), using `this` in it, and you were passing one of them to a different function defined in the global scope. Javascript scoping is complex, if you console log this inside the original close function in the first example, you can see that, when pressing esc, it was printing out window rather then the literal object you were expecting so your `this` was different. 

Alpine gives you the tools to make the structure simpler encapsulating everything in a component. 
Something like (I'm typing from my mobile, sorry, there could be typos):

```html
<div x-data="{show: false}" >
   <button="open()">open</button>
   <div x-if="show" @keydown.document="handleKey($event)">
      <button="close()"></button>
   </div>
</div>
<script>
function data() {
  return {
    show: false, 
    open() {
       this.show = true, 
       // logic to focus element
    }, 
    close() {
      this.show = false
      // additional logic on close
    },
    handleKey(event) {
      // logic where you can call this.close where needed based on your event conditions 
    } 
  }
</script>
```

**Reply by SimoTod** [reply]
Btw, glad it's sorted. It was just an observation. If you are confortable with that structure, it's okay since it's your code. 
Be aware that keydown events work a bit funny on android. I've just found it out today and I'll dig into it in the next few days but the escape modifier might not work due to android not supporting event.key (TBC). If you need to support those devices, have a go and if it does not work, you might want to use the keypress without the escape modifier and put your logic inside the handler (checking both event.key and event.which or event.keyCode like the first example).

**Reply by estevanmaito** [reply]
I see now. I've debugged it before and saw that the cleanup function was being called inside window, but totally ignored that the same was occurring with `show`.

On the encapsulation, it would become repetitive, as the `trapFocus` function would be used in other places. But I'll try it to get familiar with, as it seems definitely a clearer approach.

These were my first lines with Alpine, so I don't know if even I am comfortable with this structure 😅

**Comment by estevanmaito** [comment]
Solved. It's actually `@keydown.escape`.

I removed the handler from inside the function and just call `close` from `@keydown.escape`.

**Reply by ryangjchandler** [reply]
Sorry, I'm on mobile and quite busy so it was a rushed answer.

**Reply by estevanmaito** [reply]
Oh, no problems at all! I didn't even know it existed 😅

Just pointed for future reference.

**Reply by ryangjchandler** [reply]
I've marked my original reply as edited with the correct implementation too.

**Reply by estevanmaito** [reply]
I've edited mine as well.
